File: Lab Two - Rendezvous/rendezvous.go
// Lab Two - Rendezvous Pattern Implementation
// Description: Demonstrates the rendezvous synchronization pattern using sync.Cond
//              All goroutines must reach the barrier before any can proceed

package main

import (
	"fmt"
	"math/rand/v2"
	"sync"
	"time"
)

// Global variables shared between goroutines
// Note: Using global variables is not considered an ideal practice
var (
	mutex       sync.Mutex // Protects access to grCount
	grCount     int        // Counts goroutines that have reached the barrier
	threadCount int        // Total number of goroutines
	cond        *sync.Cond // Condition variable for barrier synchronization
)

// WorkWithRendezvous demonstrates the rendezvous pattern
// Parameters:
//   - wg: WaitGroup to signal completion
//   - Num: Goroutine identifier for output
//
// Returns:
//   - bool: Always true (success indicator)
func WorkWithRendezvous(wg *sync.WaitGroup, Num int) bool {
	// Simulate random work duration (0-4 seconds)
	var X time.Duration
	X = time.Duration(rand.IntN(5))
	time.Sleep(X * time.Second)
	fmt.Println("Part A", Num)

	// ==================== RENDEZVOUS BARRIER ====================
	// This section ensures all goroutines wait for each other
	mutex.Lock()
	grCount++ // Increment count of arrived goroutines

	if grCount == threadCount {
		// Last goroutine to arrive: wake all waiting goroutines
		cond.Broadcast()
	} else {
		// Not all here yet: wait for signal from last goroutine
		cond.Wait() // Automatically releases and reacquires mutex
	}
	mutex.Unlock()
	// ============================================================

	// All goroutines have passed the barrier
	fmt.Println("PartB", Num)
	wg.Done() // Signal completion to WaitGroup
	return true
}

// main sets up and runs the rendezvous demonstration
func main() {
	var wg sync.WaitGroup
	threadCount = 5             // Number of goroutines to synchronize
	cond = sync.NewCond(&mutex) // Initialize condition variable with mutex

	wg.Add(threadCount)
	// Launch all goroutines
	for N := range threadCount {
		go WorkWithRendezvous(&wg, N)
	}

	wg.Wait() // Wait for all goroutines to complete
}

File: Lab Two - Rendezvous/labTwo/rendezvous.cpp
#include "Semaphore.h"
#include <thread>
#include <vector>
#include <iostream>

/*! displays the first function in the barrier being executed */
void task(std::shared_ptr<Semaphore> mutexSem,std::shared_ptr<Semaphore> barrierSem, int threadCount){
  // Shared counter for threads that have reached the barrier
  static int count = 0; 
  
  std::cout << "first " << std::endl;
  
  //Barrier code - Rendezvous pattern

  // Lock the mutex to safely increment counter
  mutexSem->Wait();
  // Increment the count of threads that reached the barrier
  count++;
  // Unlock the mutex
  mutexSem->Signal(); 
  
  if(count == threadCount){
    // Last thread to arrive signals all waiting threads
    barrierSem->Signal();
  }
  
  // Wait for the signal from the last thread
  barrierSem->Wait(); 
  // Pass the signal to other waiting threads (turnstile pattern)
  barrierSem->Signal(); 
  
  std::cout << "second" << std::endl;
}

int main(void){
  std::shared_ptr<Semaphore> mutexSem;
  std::shared_ptr<Semaphore> barrierSem;
  int threadCount = 5;
  mutexSem=std::make_shared<Semaphore>(1);
  barrierSem=std::make_shared<Semaphore>(0);
  /*!< An array of threads*/
  std::vector<std::thread> threadArray(threadCount);
  /*!< Pointer to barrier object*/

  for(int i=0; i < threadArray.size(); i++){
    threadArray[i]=std::thread(task,mutexSem,barrierSem,threadCount);
  }

  for(int i = 0; i < threadArray.size(); i++){
    threadArray[i].join();
  }
  
  return 0;
}
